File: january2025/january27.py
#User function Template for python3

# design the class in the most optimal way
from collections import OrderedDict

class LRUCache:
      
    #Constructor for initializing the cache capacity with the given value.  
    def __init__(self, cap):
        #code here
        self.cap=cap
        self.cache=OrderedDict()
        
        
    #Function to return value corresponding to the key.
    def get(self, key):
        #code here
        if key not in self.cache:
            return -1
        self.cache.move_to_end(key)
        return self.cache[key]
        
        
    #Function for storing key-value pair.   
    def put(self, key, value):
        #code here
        if key in self.cache:
            self.cache.move_to_end(key)
        self.cache[key]=value
        if len(self.cache)>self.cap:
            self.cache.popitem(last=False)




#{ 
 # Driver Code Starts
#Initial Template for Python 3


def inputLine():
    return input().strip().split()


if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        capacity = int(input())
        cache = LRUCache(capacity)

        queries = int(input())
        for __ in range(queries):
            vec = inputLine()
            if vec[0] == "PUT":
                key = int(vec[1])
                value = int(vec[2])
                cache.put(key, value)
            else:
                key = int(vec[1])
                print(cache.get(key), end=" ")
        print()
        print("~")

# } Driver Code Ends

File: january2025/january27.java
import java.util.LinkedHashMap;
import java.util.Map;

class LRUCache {
    private final int cap;
    private final Map<Integer, Integer> cache;

    // Constructor for initializing the cache capacity with the given value.
    public LRUCache(int cap) {
        this.cap = cap;
        this.cache = new LinkedHashMap<Integer, Integer>(cap, 0.75f, true) {
            protected boolean removeEldestEntry(Map.Entry<Integer, Integer> eldest) {
                return size() > LRUCache.this.cap;
            }
        };
    }

    // Function to return value corresponding to the key.
    public int get(int key) {
        return cache.getOrDefault(key, -1);
    }

    // Function for storing key-value pair.
    public void put(int key, int value) {
        cache.put(key, value);
    }
}

// Driver Code
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        while (t-- > 0) {
            int capacity = sc.nextInt();
            LRUCache cache = new LRUCache(capacity);
            int queries = sc.nextInt();
            while (queries-- > 0) {
                String command = sc.next();
                int key = sc.nextInt();
                if (command.equals("PUT")) {
                    int value = sc.nextInt();
                    cache.put(key, value);
                } else {
                    System.out.print(cache.get(key) + " ");
                }
            }
            System.out.println();
            System.out.println("~");
        }
        sc.close();
    }
}

